File: docs/api-audit/probes/Tensor_07_typedevice.py
"""Behavioral probe for the Tensor audit, category 07 (dtype / device
conversion), verified against the installed cytnx==1.1.0 wheel (NOT
source-inferred).

Every runtime claim in docs/api-audit/Tensor/07-type-device-conversion.md is
backed by a report(...) assertion here. Members covered: the type/device
converters `astype` / `to` / `to_`, the deep-copy `clone`, and the two leaked
raw plumbing shims `astype_different_dtype` / `to_different_device`.

Binding-fidelity headline (T-T1, cross-ref UniTensor UT-T1): `astype`/`to`
SHORT-CIRCUIT to `is self` on a no-op conversion — a Python identity introduced
by the conti.py wrappers (cytnx/Tensor_conti.py:36-48: `if self.dtype()==dtype:
return self` / `if self.device()==device: return self`). Raw C++
`astype`/`to` return a FRESH object every call; that raw-C++ side is verified by
probes/cpp/Tensor_07_typedevice.cpp.

Two of the brief's assumptions are CORRECTED here against runtime truth (cat 04
taught us the brief can be wrong):
  * `to_` returns **None**, NOT self — C++ `void to_(const int&)` (Tensor.hpp:683)
    is bound directly (tensor_py.cpp:177), so unlike UniTensor's `to_` (which
    returns `UniTensor&`) Tensor's returns None. An in-place-return asymmetry
    (T-T4), sibling to the cat-04 `flatten_`-returns-None finding.
  * `to_`'s parameter is NAMED `device`, NOT erased — it is bound WITH
    `py::arg("device")` (tensor_py.cpp:177), so `to_(device=...)` works and
    `to_(arg0=...)` fails. This is the POSITIVE contrast to UniTensor's UT-T3
    (whose `to_` param was erased to `arg0`).

Run: source tools/env.sh && $PY docs/api-audit/probes/Tensor_07_typedevice.py
"""
import sys, os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "tools"))
import cytnx
from probe_helper import report, returns_view

Tensor = cytnx.Tensor
Type = cytnx.Type
Device = cytnx.Device

report("wheel under test is 1.1.0 (runtime, not source-inferred)",
       cytnx.__version__ == "1.1.0")


def _mat():
    """A fresh 2x2 Double Tensor of ones."""
    return cytnx.ones((2, 2), Type.Double)


# =========================================================================
# T-T1: astype / to SHORT-CIRCUIT to `is self` on a no-op conversion.
# This Python identity is BINDING-INTRODUCED by cytnx/Tensor_conti.py:36-48
# (`if self.dtype()==dtype: return self` / `if self.device()==device: return
# self`). Raw C++ astype/to return a FRESH object every call — see
# probes/cpp/Tensor_07_typedevice.cpp. Cross-ref UniTensor UT-T1.
# =========================================================================
t = _mat()
report("`astype(same dtype)` returns `self` (is self) -- a no-op short-circuit "
       "added by the conti.py wrapper, NOT raw-C++ behavior (T-T1; xref UT-T1)",
       t.astype(t.dtype()) is t)
report("`to(same device)` returns `self` (is self) -- the device no-op "
       "short-circuit from the conti.py wrapper. On this CPU-only wheel `to(cpu)` "
       "is ALWAYS the no-op path (T-T1; xref UT-T1)",
       t.to(t.device()) is t and t.to(Device.cpu) is t)

# =========================================================================
# T-T2: astype to a DIFFERENT dtype -> a DISTINCT, independent tensor of the
# requested dtype (data copied; NOT same_data).
# =========================================================================
t = _mat()
a = t.astype(Type.ComplexDouble)
report("`astype(different dtype)` returns a DISTINCT object of the requested "
       "dtype (ComplexDouble) whose data is COPIED (not same_data) -- the real "
       "conversion path (T-T2)",
       a is not t and a.dtype() == Type.ComplexDouble and not t.same_data(a))

# =========================================================================
# T-T3: clone is a DEEP copy -- an independent tensor, not a view.
# =========================================================================
report("`clone()` returns an INDEPENDENT deep copy: mutating the clone is NOT "
       "visible through the source (returns_view -> False) (T-T3)",
       returns_view(
           make=_mat,
           derive=lambda s: s.clone(),
           mutate=lambda h: h.__setitem__((0, 0), 99.0),
           read=lambda s: s[0, 0].item(),
       ) is False)
t = _mat()
c = t.clone()
report("`clone()` does NOT share storage with the source (not same_data) (T-T3)",
       c is not t and not t.same_data(c))

# =========================================================================
# T-T4: `to_` is an IN-PLACE device move that returns **None**, NOT self.
# C++ `void to_(const int&)` (Tensor.hpp:683) bound directly (tensor_py.cpp:177).
# This diverges from UniTensor's `to_` (returns UniTensor&/self); it is the same
# in-place-return asymmetry as the cat-04 `flatten_`-returns-None finding.
# =========================================================================
t = _mat()
ret = t.to_(t.device())
report("`to_(device)` moves in place and returns **None** (NOT self) -- C++ "
       "`void to_` bound directly; diverges from UniTensor's self-returning `to_` "
       "(T-T4)",
       ret is None)
report("`to_` leaves the receiver on the target device (in-place move; CPU-only "
       "here, so device stays Device.cpu) (T-T4)",
       t.device() == Device.cpu)

# =========================================================================
# T-T5: `to_`'s parameter IS named `device` -- bound WITH py::arg("device")
# (tensor_py.cpp:177). The POSITIVE contrast to UniTensor UT-T3 (erased arg0).
# =========================================================================
t = _mat()
report("`to_(device=...)` works -- the parameter is correctly named `device` "
       "(py::arg present), UNLIKE UniTensor's erased `arg0` (T-T5; xref UT-T3)",
       t.to_(device=Device.cpu) is None)
t = _mat()
raised = False
try:
    t.to_(arg0=Device.cpu)
except TypeError:
    raised = True
report("`to_(arg0=...)` RAISES TypeError -- confirming the real keyword is "
       "`device`, not the pybind default `arg0` (T-T5)",
       raised)

# =========================================================================
# T-T2: the raw plumbing shims `astype_different_dtype` / `to_different_device`
# LEAK into public dir(Tensor); each HARD-ASSERTS the argument differs (raises
# on a no-op -- they exist only so the conti.py wrapper handles the no-op) and
# does the real conversion on a genuine change. Cross-ref UniTensor UT-T2.
# =========================================================================
report("the raw plumbing bindings `astype_different_dtype` / `to_different_device` "
       "both LEAK into public dir(Tensor) (T-T2; xref UT-T2)",
       "astype_different_dtype" in dir(Tensor)
       and "to_different_device" in dir(Tensor))

t = _mat()
raised_a = False
try:
    t.astype_different_dtype(t.dtype())  # no-op -> hard error
except RuntimeError:
    raised_a = True
raised_d = False
try:
    t.to_different_device(t.device())    # no-op -> hard error
except RuntimeError:
    raised_d = True
report("`astype_different_dtype(same dtype)` / `to_different_device(same device)` "
       "each RAISE RuntimeError on a no-op -- the hard assert that forces the "
       "conti.py wrapper to intercept same-dtype/same-device calls (T-T2)",
       raised_a and raised_d)
report("`astype_different_dtype(ComplexDouble)` does the REAL conversion (the "
       "path `astype` forwards to after its short-circuit) (T-T2)",
       t.astype_different_dtype(Type.ComplexDouble).dtype() == Type.ComplexDouble)

print("Tensor 07 probe ok")

File: docs/api-audit/probes/cpp/Tensor_07_typedevice.cpp
// Raw-C++ probe for the Tensor cat-07 (dtype / device conversion)
// binding-fidelity finding T-T1. Links against a locally-built libcytnx so it
// can call the C++ methods DIRECTLY, bypassing the pybind + conti.py layer —
// proving that:
//   * T-T1: raw C++ `astype(same_dtype)` and `to(same_device)` on a NO-OP
//     return a FRESH, DISTINCT Tensor object (a new wrapper over the shared
//     _impl), NOT `*this`. The pybind primitives are wrapped by
//     cytnx/Tensor_conti.py:36-48's `if self.dtype()==dtype: return self` /
//     `if self.device()==device: return self` short-circuit, so the Python
//     `astype(same) is self` / `to(same) is self` identity is
//     BINDING-INTRODUCED, not C++ behavior.
//   * A REAL conversion (astype to a different dtype) and `clone()` return
//     DISTINCT objects that do NOT share data (independent copies) — the
//     raw-C++ side of T-T2 / T-T3.
//
// NOTE on `to_` (T-T4/T-T5, no C++ assertion needed): unlike UniTensor's `to_`
// (which returns `UniTensor&`), Tensor's C++ `to_` is declared
// `void to_(const int&)` (Tensor.hpp:683) — it CANNOT return `*this`, which is
// exactly why the Python `to_` returns None (T-T4). Its Python parameter IS
// named `device` because the binding adds py::arg("device") (tensor_py.cpp:177,
// T-T5) — both are source/pybind facts, verified by the Python probe.
#include "cytnx.hpp"
#include <iostream>
#include <string>
using namespace cytnx;

static int fails = 0;
static void report(const std::string& claim, bool ok) {
  std::cout << (ok ? "[PASS] " : "[FAIL] ") << claim << std::endl;
  if (!ok) ++fails;
}

int main() {
  // -----------------------------------------------------------------------
  // T-T1: astype(same_dtype) is a NO-OP conversion. C++ returns a FRESH,
  // DISTINCT Tensor wrapper (by value) sharing the underlying _impl — it is
  // NOT `*this`. Two consecutive no-op calls yield two DISTINCT objects, both
  // sharing data with the source: proof there is no `return *this` / `is self`
  // short-circuit in C++. (The Python conti.py wrapper adds that short-circuit.)
  // -----------------------------------------------------------------------
  {
    Tensor T = ones({2, 2});                 // Double, CPU
    Tensor R1 = T.astype(T.dtype());         // no-op conversion (same dtype)
    Tensor R2 = T.astype(T.dtype());         // a second no-op call
    bool distinct = (&R1 != &T) && (&R2 != &T) && (&R1 != &R2);
    bool shared = T.same_data(R1) && T.same_data(R2);
    report("T-T1: C++ astype(same_dtype) returns a FRESH, DISTINCT Tensor object "
           "(two no-op calls give two distinct wrappers, each != the source) that "
           "SHARES storage with the source (same_data) -- there is NO "
           "`return *this`/`is self` short-circuit in C++; the Python "
           "`astype(same) is self` identity is binding-introduced (Tensor_conti.py)",
           distinct && shared);
  }

  // -----------------------------------------------------------------------
  // T-T1: to(same_device) is likewise a NO-OP. C++ returns a FRESH, DISTINCT
  // wrapper sharing the _impl -- not `*this`.
  // -----------------------------------------------------------------------
  {
    Tensor T = ones({2, 2});
    Tensor R = T.to(T.device());             // no-op (same device)
    report("T-T1: C++ to(same_device) returns a FRESH, DISTINCT Tensor object "
           "(!= the source) that SHARES storage (same_data) -- again no `is self` "
           "short-circuit; the Python `to(same) is self` identity is "
           "binding-introduced (Tensor_conti.py)",
           (&R != &T) && T.same_data(R));
  }

  // -----------------------------------------------------------------------
  // T-T2 / T-T3: a REAL conversion (astype to a DIFFERENT dtype) copies data:
  // distinct object, NOT same_data. clone() is likewise an independent deep copy.
  // -----------------------------------------------------------------------
  {
    Tensor T = ones({2, 2});                 // Double
    Tensor A = T.astype(Type.ComplexDouble);
    report("T-T2: C++ astype(different_dtype) returns a DISTINCT, INDEPENDENT "
           "object (new dtype, data copied -> NOT same_data) -- the real-conversion "
           "path the Python raw astype_different_dtype shim forwards to",
           (&A != &T) && (A.dtype() == Type.ComplexDouble) && !T.same_data(A));

    Tensor C = T.clone();
    report("T-T3: C++ clone() returns a DISTINCT, INDEPENDENT deep copy (NOT "
           "same_data with the source) -- backing the Python clone independence",
           (&C != &T) && !T.same_data(C));
  }

  std::cout << (fails ? "cpp probe FAILED" : "cpp probe ok") << std::endl;
  return fails ? 1 : 0;
}
